### ISSUE TITLE
Active User/Device Definition 

### ISSUE THREAD

**Issue by hakkiToran** [camaraproject/RegionDeviceCount#21]
**Problem description**
The current project documentation on the [Region User Count project page](https://camaraproject.org/region-user-count/) defines "active users," but the definition is unclear. Specifically, it is not clear whether "active users" include only those actively using data or if it also considers users who are connected but idle.
Please check this website :
What is the definition of "active user"? Does an "active user" include users in an IDLE state as described [here]
https://worldtechie.blogspot.com/2013/04/lte-ecmemm-and-rrc-states.html
![image](https://github.com/camaraproject/RegionUserCount/assets/171144019/c22dae57-a3f7-4157-aa3d-ef2e0a3df147)

**Possible evolution**
Provide a detailed and clear definition of "active user" within the documentation. Specify whether it includes both users actively using data and those connected but idle. 

**Alternative solution**
Include examples or scenarios illustrating the "active user" status to enhance understanding.

**Additional context**

Clarifying this definition will help users accurately understand and utilize the API, ensuring better implementation and integration.

**Comment by chinaunicomyangfan** [comment]
@hakkiToran  Hello,Welcome to join our discussion.
We have discussed the definition of active users [here](https://github.com/camaraproject/RegionUserCount/issues/8) and ultimately determined that the number counted by this API is the number of connected devices. As long as they are connected, they are counted regardless of whether they are idle. We have also decided to change the API name to Region Device Count. This has been submitted to the GSMA and apibacklog working groups

**Comment by hakkiToran** [comment]
Actually except from the definition, If you query the number of User for  time=t for, what you will get ?

You will get the non-IDLE users. However the problem is most of time the users will be on IDLE for a specific time=t. Maybe only %20 of the users will be non-IDLE. Because if you leave your phone on table, maybe after 10-15 seconds, the phone will change to IDLE state. That's why querying active users for a specific time=t gives you may be maximum %20 percent of the users. 
Most of the time, the UEs will be on IDLE state, if not actively a person is using in hand. For example in earthquake scenario, most probably, the specific time=t based queries will miss to cover more than %80 of users.
<img width="595" alt="image" src="https://github.com/camaraproject/RegionUserCount/assets/171144019/ac282df8-4c51-4bbc-84f0-2511fb396ab4">

https://worldtechie.blogspot.com/2013/04/lte-ecmemm-and-rrc-states.html
<img width="607" alt="image" src="https://github.com/camaraproject/RegionUserCount/assets/171144019/5e43ec8d-4f63-4a26-822b-e3eb42db6ce5">


 **How can wesolve this problem?** 
We need to query for a time interval instead of quering for a specific time. If time interval for example 15 minutes or 1 hour is used, the result will most probably show the more than %95 percent of user. Because UE and network making background signaling peridically like paging cycles, TAU times.., so in this time interval the phone must have made activities even in the pocket or on table . So the the time interval based query cover more users.

**Comment by gregory1g** [comment]
We have agreed that API is about devices, not users. So, could you please describe the issue in "device" terminology.

**Comment by gregory1g** [comment]
Let's change terminology to devices in this discussion. Ideally starting from the title of the issue.

**Comment by hakkiToran** [comment]
@gregory1g I changed the name as user/device.  Changing the name for this issue does not eliminate the not getting the IDLE users . 

**How can wesolve this problem?**
We need to query for a time interval instead of quering for a specific time. If time interval for example 15 minutes or 1 hour is used, the result will most probably show the more than %95 percent of user/device. Because UE and network making background signaling peridically like paging cycles, TAU times.., so in this time interval the phone must have made activities even in the pocket or on table . So the the time interval based query cover more users/devices in the region, which will be more correct.

**Comment by chinaunicomyangfan** [comment]
@hakkiToran In my understanding, as long as the device is turned on and connected to the network, regardless of its status, it needs to be counted. For idle devices, regular location update signaling can be used to count. So I believe that the number of connected devices can be counted using time points or time windows

**Comment by hakkiToran** [comment]
@chinaunicomyangfan You are right, when a time interval based query made like for example (**T**-**xx** minutes ) where xx configurable can be much more useful.  The value of XX should be for example 15 < xx < 300 minutes. Beca
One important point to consider with time  interval based (**T**-**xx** minutes )  query is to get the last cell. Because some users may be moving under different cells. So the API should return the last cell during the time interval.

**Comment by gregory1g** [comment]
@hakkiToran , this sounds like last known location in [T-xx; T] time frame. And sounds like a reasonable definition/approach to me.

**Comment by hakkiToran** [comment]
@chinaunicomyangfan and @philipxujin Will you implement this suggestions to the code ?

**Comment by chinaunicomyangfan** [comment]
@hakkiToran 
Sorry for the late response. We have been busy with the Shanghai MWC recently. We will carefully consider your proposal and add it to the code. It is expected to be completed next week

**Comment by chinaunicomyangfan** [comment]
@hakkiToran I don't know if I understand correctly. Your requirement should be the number of devices that have visited a certain area within a certain time window,any devices that have accessed this area should be included in the calculation,right?

**Comment by gregory1g** [comment]
> If you query without giving a time interval, you can not output all the devices in that region

You can. API can assume some interval by default. For example, 10 minutes. 

> 1 hour is used, the result will most probably show the much more than %95 percent of devices.
95% of what number?  
The important question we should answer here is - what the API tries to expose:
- either an average/immediate number of devices in the area (as accurate as it can)
-  or number of unique devices visited that the area during some time.

Some small area in a train station could be a good example of the difference:
there is always 8-10 people there at the same time, but people constantly move thought this area:
- the most accurate average/immediate number is 10
- but if we count unique devices per hour, the answer can be 1000.

**Comment by hakkiToran** [comment]
@gregory1g 

10 minutes by default is also an interval. In addition to that the time interval should be configurable and the interval should be for example extentable from 15 minutes to 6 or 8 hours. The unique device count should be calculated. 

Different scenarios may need different time intervals, for example, at night users may be sleeping and there may not be an activity for a long time. For earthquake, the users may not be making any activity, however we need to count them.

Another important point to consider with time interval based (T-xx minutes ) query is to get the last cell. Because some users may be moving under different cells. So the API should return the last cell during the time interval.

**Comment by gregory1g** [comment]
> 10 minutes by default is also an interval. 
sure, I just say it that the interval can be an either a implementation details or exposed via the API. 

>Different scenarios may need different time intervals, for example, at night users may be sleeping and there may not be an activity for a long time. For earthquake, the users may not be making any activity, however we need to count them.

Good example. If the API will return all devices which have been seen in the area during this "long time" we will report devices which already left the area and are connected to other cells. For an train station this could be 100 times bigger number. Should the client guess how static the dynamic the crowd is in the target area?
In other words: does API returns number of devices in the area (as accurate as it can) or number of devices visited the area?

Can you please clarify what are realistic time intervals when a phone completely idles?

**Comment by hakkiToran** [comment]
@gregory1g 

It is better to have a configurable time interval. Because different scenarios may need different interval requirement.

You can think for the time durations as below.
<img width="608" alt="image" src="https://github.com/camaraproject/RegionDeviceCount/assets/171144019/22bad89c-9358-48ba-9c49-50e32fa000ed">

Another important point to consider with time interval based (T-xx minutes ) query is to get the last cell. Because some users may be moving under different cells. So the API should return the last cell during the time interval.

**Comment by chinaunicomyangfan** [comment]
I think we can add a startTime parameter and an endTime parameter to the request parameters, both of which are optional parameters. If the user has configured these two parameters, the devices that have accessed this area during this period will be counted. The following points may need to be discussed
1. The startTime cannot be too early, it is recommended not to be earlier than 7 days from the current time
2.Is there a need to limit the time between startTime and endTime? It is recommended that it should not exceed 24 hours
3. startTime and endTime cannot be transmitted separately
4. If both startTime and endTime are not entered, in actual calculation, considering that devices in idle state cannot be counted, a default time interval needs to be pushed forward from the current time, which is recommended to be 10 minutes

**Comment by hakkiToran** [comment]
@chinaunicomyangfan 

I am agree with your comments. In addition to your comments, I added another point to consider.

1. If a device seen in different cells in the queried time interval (Endtime- StartTime), the last cell of the device should be returned . The device should be counted under the last seen cell.

**Comment by chinaunicomyangfan** [comment]
@hakkiToran @gregory1g I have added two request parameters **starttime** and **endtime** in YAML file. Please review and provide improvement suggestions. Thank you

**Comment by hakkiToran** [comment]
@chinaunicomyangfan, Thank you for the update information. I will also check it.

**Comment by hakkiToran** [comment]
@chinaunicomyangfan Thank you for the change. I checked it. It is OK for me.
